raft.go

```diff
@@ -346,49 +353,78 @@ func (r *Raft) runLeader() {
-	inConfig := make(map[ServerID]bool, len(r.configurations.latest.Servers))
-	lastIdx := r.getLastIndex()
+// updatePeeers will start communication with new peers, send new and
+// existing peers updated control information, and stop communication with
+// removed peers. This must only be called from the main thread.
+func (r *Raft) updatePeers() {
+	inConfig := make(map[ServerID]Server, len(r.configurations.latest.Servers))
 
 	// Start replication goroutines that need starting
 	for _, server := range r.configurations.latest.Servers {
 		if server.ID == r.localID {
 			continue
 		}
-		inConfig[server.ID] = true
-		if _, ok := r.leaderState.replState[server.ID]; !ok {
+		inConfig[server.ID] = server
+		if _, ok := r.peers[server.ID]; !ok {
 			r.logger.Printf("[INFO] raft: Added peer %v, starting replication", server.ID)
-			s := &followerReplication{
-				peer:        server,
-				commitment:  r.leaderState.commitment,
-				stopCh:      make(chan uint64, 1),
-				triggerCh:   make(chan struct{}, 1),
-				currentTerm: r.getCurrentTerm(),
-				nextIndex:   lastIdx + 1,
-				lastContact: time.Now(),
-				notifyCh:    make(chan struct{}, 1),
-				stepDown:    r.leaderState.stepDown,
+			controlCh := startPeer(server.ID, server.Address, r.logs, r.snapshots,
+				r.goRoutines, r.trans, r.localAddr, r.peerProgressCh, peerOptions{
+					logger:            r.logger,
+					maxAppendEntries:  r.conf.MaxAppendEntries,
+					heartbeatInterval: r.conf.HeartbeatTimeout / 5,
+				})
+			peer := &raftPeer{
+				controlCh: controlCh,
 			}
-			r.leaderState.replState[server.ID] = s
-			r.goFunc(func() { r.replicate(s) })
-			asyncNotifyCh(s.triggerCh)
+			r.peers[server.ID] = peer
 		}
 	}
 
-	// Stop replication goroutines that need stopping
-	for serverID, repl := range r.leaderState.replState {
-		if inConfig[serverID] {
-			continue
+	// Send new control information and stop Peer goroutines that need stopping
+	lastIndex, lastTerm := r.getLastEntry()
+	for serverID, peer := range r.peers {
+		role := r.getState()
+		if role == Shutdown {
+			// Shutdown must be the last control signal sent to the Peer, so we mask
+			// it here and send it later in shutdownPeers().
+			role = Follower
+		}
+		verifyCounter := r.verifyCounter
+		server, ok := inConfig[serverID]
+		if !ok {
+			r.logger.Printf("[INFO] raft: Removed peer %v, stopping communication", serverID)
+			delete(r.peers, serverID)
+			role = Shutdown
+		} else {
+			if server.Suffrage != Voter {
+				if role == Candidate {
+					role = Follower
+				}
+				verifyCounter = 0
+			}
+		}
+		control := peerControl{
+			term:          r.getCurrentTerm(),
+			role:          role,
+			verifyCounter: verifyCounter,
+			lastIndex:     lastIndex,
+			lastTerm:      lastTerm,
+			commitIndex:   r.getCommitIndex(),
 		}
-		// Replicate up to lastIdx and stop
-		r.logger.Printf("[INFO] raft: Removed peer %v, stopping replication after %v", serverID, lastIdx)
-		repl.stopCh <- lastIdx
```

Comment:
The new code makes no attempt to continue replicating entries to a removed server. With the membership changes the way they are now, that didn't actually cause any benefit AFAIK. Let's talk about what we need to happen here.